src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+ * \param[in]   shard_md	Shard metadata.
+ * \param[out]  layout_pp       The layout generated for the object. Contains
+ *                              references to the targets in the pool map where
+ *                              the shards will be placed.
+ *
+ * \return                      An integer value containing the error
+ *                              code or 0 if the function returned
+ *                              successfully.
+ */
+static int
+mapless_obj_place(struct pl_map *map, struct daos_obj_md *md,
+		  struct daos_obj_shard_md *shard_md,
+		  struct pl_obj_layout **layout_pp)
+{
+	struct pl_mapless_map           *mmap;
+	struct pl_obj_layout            *layout;
+	struct mpls_obj_placement    mop;
+	d_list_t                        remap_list;
+	daos_obj_id_t                   oid;
+	int                             rc;
+
+	mmap = pl_map2mplmap(map);
+	oid = md->omd_id;
+
+	rc = mpls_obj_placement_get(mmap, md, shard_md, &mop);
+	if (rc) {
+		D_ERROR("mpls_obj_placement_get failed, rc %d.\n", rc);
+		return rc;
+	}
+
+	/* Allocate space to hold the layout */
+	rc = pl_obj_layout_alloc(mop.mop_grp_nr * mop.mop_grp_size, &layout);
+	if (rc != 0) {
+		D_ERROR("pl_obj_layout_alloc failed, rc %d.\n", rc);
+		return rc;
+	}
+
+	/* Get root node of pool map */
+	D_INIT_LIST_HEAD(&remap_list);
+	rc = get_object_layout(mmap, layout, &mop, &remap_list, md);
+	if (rc < 0) {
+		D_ERROR("Could not generate placement layout, rc %d.\n", rc);
+		pl_obj_layout_free(layout);
+		mapless_remap_free_all(&remap_list);
+		return rc;
+	}
+
+	*layout_pp = layout;
+	mapless_obj_layout_dump(oid, layout);
+
+	mapless_remap_free_all(&remap_list);
+	return DER_SUCCESS;
+}
+
+/**
+ *
+ * \param[in]   map             The placement map to be used to generate the
+ *                              placement layouts and to calculate rebuild
+ *                              targets.
+ * \param[in]   md              Metadata describing the object.
+ * \param[in]   shard_md        Metadata describing how the shards.
+ * \param[in]   rebuild_ver     Current Rebuild version
+ * \param[out]   tgt_rank       The rank of the targets that need to be rebuild
+ *                              will be stored in this array to be passed out
+ *                              (this is allocated by the caller)
+ * \param[out]   shard_id       The shard IDs of the shards that need to be
+ *                              rebuilt (This is allocated by the caller)
+ * \param[in]   array_size      The max size of the passed in arrays to store
+ *                              info about the shards that need to be rebuilt.
+ * \param[in]	myrank		The rank of the server. Only servers who are
+ *				the leader for a particular failed shard will
+ *				initiate a rebuild for it.
+ *
+ * \return			The number of shards that need to be rebuilt on
+ *				another target, Or 0 if none need to be rebuilt.
+ */
+static int
+mapless_obj_find_rebuild(struct pl_map *map, struct daos_obj_md *md,
```

Comment:
Dup of ring_obj_find_rebuild()?

Reply:
Required implementation of Rebuild API
".o_obj_find_rebuild     = mapless_obj_find_rebuild,"

Factored out code related to DTX.